include/watchdog.h

```diff
@@ -1,116 +1,206 @@
+ * the watchdog will trigger.
+ *
+ * @param up Upper limit of watchdog reload timeout in miliseconds.
+ * @param down Lower limit of watchdog reload timeout in miliseconds.
+ */
+struct wdt_window {
+	u16_t up;
+	u16_t down;
 };
 
 /**
- * WDT clock cycles for timeout type.
+ * @brief Watchdog configuration struct.
+ *
+ * @param options - a bitfield with following parts:
+ *    enable callback     [ 0 ]       - reset or callback -> reset
+ *    pause in sleep      [ 1 ]       - pause in CPU sleep
+ *    pause by debugger   [ 2 ]       - pause when halted by debugger
  */
-enum wdt_clock_timeout_cycles {
-	WDT_2_16_CYCLES,
-	WDT_2_17_CYCLES,
-	WDT_2_18_CYCLES,
-	WDT_2_19_CYCLES,
-	WDT_2_20_CYCLES,
-	WDT_2_21_CYCLES,
-	WDT_2_22_CYCLES,
-	WDT_2_23_CYCLES,
-	WDT_2_24_CYCLES,
-	WDT_2_25_CYCLES,
-	WDT_2_26_CYCLES,
-	WDT_2_27_CYCLES,
-	WDT_2_28_CYCLES,
-	WDT_2_29_CYCLES,
-	WDT_2_30_CYCLES,
-	WDT_2_31_CYCLES
+struct wdt_cfg {
+	u8_t options;
 };
 
+/** Watchdog callback */
+typedef void (*wdt_callback_t)(struct device *dev, void *handle);
 
 /**
- * WDT configuration struct.
+ * @brief Watchdog timeout configuration struct.
+ *
+ * @param window - timing parameters of watchdog timeout.
+ * @param handle - unique handle used to reload specific watchdog timeout.
+ *		   This handle will also be supplied to callback if
+ *		   WDT_MODE_CALLBACK is set.
+ * @param callback - timeout callback.
+ * @param next - pointer to next timeout configuration. It is used for
+ *		 watchdogs with staged timeouts functionality. Should be NULL
+ *		 for single stage timeout.
  */
-struct wdt_config {
-	u32_t timeout;
-	enum wdt_mode mode;
-	void (*interrupt_fn)(struct device *dev);
+struct wdt_timeout_cfg {
+	wdt_window window;
+	void *handle;
+	wdt_callback_t callback;
+	struct wdt_timeout_cfg *next;
 };
 
-typedef void (*wdt_api_enable)(struct device *dev);
-typedef void (*wdt_api_disable)(struct device *dev);
-typedef int (*wdt_api_set_config)(struct device *dev,
-				  struct wdt_config *config);
-typedef void (*wdt_api_get_config)(struct device *dev,
-				   struct wdt_config *config);
-typedef void (*wdt_api_reload)(struct device *dev);
-
 struct wdt_driver_api {
-	wdt_api_enable enable;
-	wdt_api_disable disable;
-	wdt_api_get_config get_config;
-	wdt_api_set_config set_config;
-	wdt_api_reload reload;
+	int (*start)(struct device *dev);
+	int (*stop)(struct device *dev);
+	int (*configure)(struct device *dev, struct wdt_cfg *cfg);
+	int (*install_timeout)(struct device *dev, struct wdt_timeout_cfg *cfg);
+	int (*reload)(struct device *dev, void *handle);
 };
 
-static inline void wdt_enable(struct device *dev)
+/**
+ * @brief Enable watchdog instance.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 if successful.
+ * @retval -EFAULT if watchdog instance has not been configured yet.
+ * @retval -EBUSY if watchdog instance has been already started.
+ */
+static inline int wdt_start(struct device *dev)
```

Comment:
This function is Nordic nRF51 specific. Typically a watchdog is started/stopped when it is configured. Also virtually all watchdogs are started upon reset so having an explicit wdt_start function is confusing. Watchdogs can only be reconfigured. The function can be removed.

Reply:
hum depends actually. On some SoC you can really stop wdt running via clock controller, and make it running again the same way. That does not seem nrf specific to me.

Reply:
K64 subfamily can be explicitly stopped through the WDOG_EN bit in the watchdog status and
control register. nRF5 family can be explicitly started but can't be explicitly stopped. I would left both `wdt_start` and `wdt_stop` and return -ENOTSUP if SoC does not support called functionality.

Reply:
@tbursztyka, @m-kru you are both right and I understand the reasoning behind having the two functions but I still would like to argue that we are better without them.

There are some watchdogs which can be stopped and then restarted but this is a niche behavior. If we get rid of wdt_start we will still support watchdogs which allow to be stopped and restarted with wdt_stop, wdt_configure combo. Rather than having
```
wdt_stop()
wdt_install_timeout()
...
wdt_install_timeout()
wdt_configure()
wdt_start()
```
we would have
```
wdt_stop()
wdt_install_timeout()
...
wdt_install_timeout()
wdt_configure()
```
wdt_configure() and wdt_start() would have to be used in pairs anyway. 

When writing an application we generally shouldn't concern ourselves with the exact driver implementation but the way API was designed. So we would need to use wdt_configure() and wdt_start() also for watchdogs that can't be re-started.

To sum up:
- We don't loose any functionality if we remove this function. 
- For majority of the users familiar with the way their watchdog works the API will be more straightforward to use.

Reply:
@mnkp Would `wdt_configure()` automatically start watchdog or should it be configurable (I think it should automatically start it as there is rather no sense in configuring watchdog and not enabling it)?

Btw. I have realized that in API RFC there is no word about uninstalling timeouts. Should we provide such functionality via separate function, or maybe `wdt_stop()` should automatically clear all installed timeouts in addition to stopping watchdog? (If yes I think it should be renamed to `wdt_disable()`).

Reply:
> Would `wdt_configure()` automatically start watchdog or should it be configurable 

Yes, in this case `wdt_configure()` should automatically start the watchdog. This is in fact how most watchdog modules are implemented. Writing the configuration register will automatically re-start the module.

> Should we provide such functionality via separate function, or maybe `wdt_stop()` should automatically clear all installed timeouts in addition to stopping watchdog?

Yes, I believe it's better if `wdt_stop()` automatically clears all installed timeouts. If we stop and re-start the watchdog it's because we want to re-configure it. Otherwise we should keep feeding it. The name 'wdt_disable` would indeed be more appropriate here.

Reply:
wdt_configure() does not start the watchdog, it installs the controller's "behavior".

about loosing installed timeouts, I guess it depends on controllers again.

Reply:
We need only one function to configure and start the watchdog. Maybe we can come up with a different name? The only alternative that comes to my mind is wdt_control.

Reply:
`wdt_setup()` ?
@tbursztyka @mnkp Is there any agreement if there should be one or two functions for configuring and starting watchdog, cause I am a bit lost with that issue right now?

Reply:
well, maybe mixing up controller's config and starting is a good idea. wdt_start() was not bad, just remove wdt_config and put the parameter into wdt_start?

Reply:
wdt_start() is not a good name for this function as most of the watchdog modules will already be running when the function is called. User seeing wdt_start() function and not calling it will be right to assume that the watchdog will stay in the disabled state. That's incorrect. 

I think `wdt_setup()` is a really good name as it reflects well what the function is doing.